lib/active_model/serializer/adapter.rb

```diff
@@ -101,43 +102,12 @@ def fragment_cache(*args)
         raise NotImplementedError, 'This is an abstract method. Should be implemented at the concrete adapter.'
       end
 
-      private
-
       def cache_check(serializer)
-        @cached_serializer = serializer
-        @klass             = @cached_serializer.class
-        if is_cached?
-          @klass._cache.fetch(cache_key, @klass._cache_options) do
-            yield
-          end
-        elsif is_fragment_cached?
-          FragmentCache.new(self, @cached_serializer, @options).fetch
-        else
+        CachedSerializer.new(serializer).cache_check(self) do
```

Comment:
Extraction of latent class not required for PR.  All the other methods are essentially private methods in the service of `cache_check` that only work if cache_check is first called to set the ivars